util/dvsim/FpvCfg.py

```diff
@@ -0,0 +1,244 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+from pathlib import Path
+
+import hjson
+from tabulate import tabulate
+
+from OneShotCfg import OneShotCfg
+from utils import subst_wildcards
+
+
+class FpvCfg(OneShotCfg):
+    """Derivative class for FPV purposes.
+    """
+    def __init__(self, flow_cfg_file, proj_root, args):
+        super().__init__(flow_cfg_file, proj_root, args)
+        self.header = ["name", "proven", "cex", "undetermined", "covered", "unreachable",
+                       "pass_rate", "cov_rate"]
+        self.summary_header = ["name", "pass_rate", "stimuli_cov", "coi_cov", "prove_cov"]
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.results_title = self.name.upper() + " FPV Results"
+
+    def parse_dict_to_str(self, input_dict):
+        # '''
+        # This is a helper function to parse dictionary items into a string
+        # Example Input:
+        # {
+        #   "unreachable": ["prop1, prop2, prop3"],
+        #   "cex"        : ["prop1"],
+        # }
+        # Example Output:
+        # string = "unreachable:
+        # prop1
+        # prop2
+        # prop3
+        # cex:
+        # prop1"
+        # '''
+        output_str = ""
+        for key, item in input_dict.items():
+            if item:
+                output_str += "\n" + key + ":\n"
+                output_str += "\n".join(item)
+        return output_str
+
+    def gen_results_summary(self):
+        # '''
+        # Gathers the aggregated results from all sub configs
+        # The results_summary will only contain the passing rate and
+        # percentages of the stimuli, coi, and proven coverage
+        # The email_summary will contain all the information from results_md
+        # '''
+        log.info("Create summary of FPV results")
+
+        results_str = "## " + self.results_title + " (Summary)\n\n"
+        results_str += "### " + self.timestamp_long + "\n\n"
+
+        colalign = ("center", ) * len(self.summary_header)
+        table = [self.summary_header]
+        for cfg in self.cfgs:
+            try:
+                table.append(cfg.result_summary[cfg.name])
+            except Exception as e:
+                table.append([cfg.name, "ERROR", "N/A", "N/A", "N/A"])
+                log.error("cfg: %s could not find generated results_summary: %s", cfg.name, e)
+        if len(table) > 1:
+            self.results_summary_md = results_str + tabulate(
+                table, headers="firstrow", tablefmt="pipe", colalign=colalign)
+        else:
+            self.results_summary_md = results_str
+
+        log.info("[result summary]: %s", self.results_summary_md)
+
+        # Generate email results summary
+        colalign = ("left", ) + ("center", ) * (len(self.header) - 1)
+        email_table = [self.header]
+        error_message = ""
+
+        for cfg in self.cfgs:
+            email_result = cfg.result.get("fpv_summary")
+            if email_result is not None:
```

Comment:
i see, thanks Michael, updated

Reply:
I'm really sorry, but that's not quite right (because the empty string is falsy in Python). Maybe it doesn't matter here, but if the aim is to check whether the key occurs in the dictionary, you need an identity comparison with None.